### DISCUSSION TITLE
audio / video sync issues

### DISCUSSION THREAD

**Discussion by GitBigDaddy** [paullouisageneau/libdatachannel#727] [Q&A]
Greetings, 

The system I am working on is based off of the 'streamer' example. We have several possible sources that are combined into a single outgoing AV stream that's in-sync prior to "track->send(). Audio and video quality are good however occasional gaps in the video frame cadence appear to produce delays (small at first) between audio and video. Seems to be cumulative. After about 5 minutes of video we have seen delays in the 5-10 second range. 

I have attempted to address the issue prior to sending into the "tracks" (by adjusting timestamps) and I am just missing something. How does one develop an approach to keep channels in sync. Thanks!

**Comment by mondain** [comment]
I've been working on some timestamp logic and it seems to work properly as far as I can tell. I see the initial post is quite old and I don't know how long I'll keep the branch, but if you see this and it helps then awesome. Here's were we get audio and video from the encoders https://github.com/mondain/obs-studio-whip1/blob/temp/plugins/obs-webrtc/whip-output.cpp#L72 we're using DTS in microseconds so I immediately divide by 1000 so we'll be in the milliseconds range and multiply by the rate by type. Then we hit Send here: https://github.com/mondain/obs-studio-whip1/blob/temp/plugins/obs-webrtc/whip-output.cpp#L559

Contents of `Data` and `Send` for posterity here:

```c++
void WHIPOutput::Data(struct encoder_packet *packet)
{
	if (packet) {
		// don't send media unless our peer is connected
		if (peer_connection != -1) {
			// reduce to milliseconds from microseconds (0 based dts counter)
			uint64_t dts_ms = (packet->dts_usec / 1000);
			if (packet->type == OBS_ENCODER_AUDIO) {
				uint64_t frame_ts =
					(dts_ms - last_audio_dts_ms) * 48;
				/*
				do_log(LOG_INFO, "Audio: %u frame ts: %u",
				       dts_ms, frame_ts);
				*/
				Send(audio_track, packet->data, packet->size,
				     frame_ts);
				last_audio_dts_ms = dts_ms;
			} else if (packet->type == OBS_ENCODER_VIDEO) {
				uint64_t frame_ts =
					(dts_ms - last_video_dts_ms) * 90;
				/*
				do_log(LOG_INFO, "Video: %u frame ts: %u",
				       dts_ms, frame_ts);
				*/
				Send(video_track, packet->data, packet->size,
				     frame_ts);
				last_video_dts_ms = dts_ms;
			}
		}
	} else {
		Stop(false);
		obs_output_signal_stop(output, OBS_OUTPUT_ENCODE_ERROR);
	}
}

void WHIPOutput::Send(int track, void *data, uintptr_t size, uint64_t ts)
{
	// get the current rtp timestamp for the track
	uint32_t current_timestamp = 0;
	rtcGetCurrentTrackTimestamp(track, &current_timestamp);
	// calculate the timestamp using the media clock rate, plus current
	uint64_t rtp_timestamp = current_timestamp + ts;
	// ensure that the timestamp is within the valid range
	if (rtp_timestamp > 0xFFFFFFFF) {
		rtp_timestamp = rtp_timestamp % 0xFFFFFFFF;
	}
	do_log(LOG_INFO, "Send ts: %u calculated rtp ts: %u", ts,
	       rtp_timestamp);
	// set new timestamp
	rtcSetTrackRtpTimestamp(track, static_cast<uint32_t>(rtp_timestamp));
	// send the track data
	rtcSendMessage(track, reinterpret_cast<const char *>(data), (int)size);
	// update total after send completes
	total_bytes_sent += size;
}
```

Hope it helps someone and if my logic is incorrect, feel free to let me know.
